chap-cheri-riscv.tex

```diff
@@ -1448,3 +1343,62 @@ \subsection{\PCC{} alignment}
 % \asm{SW}.
 
 \pmnote{Need a section on capability state on reset.}
+
+\section{Design Rationale}
+
+The CHERI-RISC-V specification has evolved over time.  This section
+describes some of the altenatives considered and other changes made.
+
+\subsection{Separate Capability Register File}
+
+CHERI-RISC-V was originally specified with two different approaches
+for adding general-purpose capability registers.  One approach
+described earlier extended existing integer general-purpose registers
+to hold capabilities (a ``merged'' register file).  The second
+approach left the integer register file as-is and added a separate
+capability register file to hold capabilities (a ``split'' register
+file).  This second approach mirrored the approach used on CHERI-MIPS.
+
+Most of the ISA is agnostic to this choice due to the principle of
+intentional-use such as the explicit use of capability-relative
+addresses versus integer pointers relative to \DDC{} or \PCC{}.
+However, the ``split'' register file approach did require a few
+additional changes:
+
+\begin{itemize}
```

Comment:
CSub and CGetAddr are also not needed in a merged world, though currently we provide them

Reply:
Hmm, I can document them here as still existing, but that is true, and I think the x86 sketch doesn't provide either of these.

Reply:
This does open the question of if we want to deprecate these instructions?  Clear I Think wasn't actually implemented on anything so it is simpler to just outright remove.

Reply:
`CGetAddr` goes away only if we assume that merged register files hold the address (and only the address) in their "integer subregisters".  See https://github.com/CTSRD-CHERI/cheri-specification/wiki/Tracking-discussion-for-CGetAddr-vs-CToPtr-vs-as-integer-alias-vs-CGetLow for more discussion. 

`CSub` can probably go away in the merged world regardless of layout issues.

Reply:
True, but we currently don't yet consider merged register files where that isn't true.  Morello still follows the TBI model.  I'm not sure if the MTE-like thing for CHERI-RISC-V also follows that model?  Regardless, if we decide to amend the spec to permit registers where CGetAddr isn't equivalent to CGetLow, that change can update this text.

Reply:
> I'm not sure if the MTE-like thing for CHERI-RISC-V also follows that model?

The version I have stole some otype bits to store a 4-bit version. I didn't want to get into the complication of using address bits for this, although it might be preferable in the long run.

Reply:
I meant to reply again to this yesterday.  Maybe we could soften the language around `CGetAddr` to suggest that it is not currently needed but might be needed if we stored additional data other than the address in the aliased integer register in the future?

Reply:
(Replying over here from the other sub-thread): I've dropped the language update for `CSub`, but have kept the note about possibly needing `CGetAddr` in the future still.  @nwf is this updated version better?

Reply:
Seems fine.  I think `CSub` needs a similar note, though, for the case that there's more than just the address in the aliased integer sub-register:
- I think we'd want it that `CSub` really does just sign-extend the subtraction of the addresses while `Sub` just projects the lower words and subtracts those, but
- I guess there's the alternative where `sub` is tag-dependent and subtracting two registers with both tags set is different than anything else

Reply:
I originally had a note for `CSub`, but @jrtc27 pointed out that pointer subtraction is only well-defined for subtracting between pointers to the same object.  Those pointers would have the same MTE-type color, so plain `Sub` would still work fine.  Based on that reason I removed the comment from `CSub`.  I don't think `Sub` should be tag-dependent as we've generally avoided having tag-dependent behavior like this.

Reply:
OK, that's probably fair as far as the current architecture document goes.  As a completionist nit, tho', it's possible that there could be other things than MTE colors in the non-address bits in the lower word; additional permission bits, for example, and those _are_, one assumes, allowed to vary between pointers to the same object.

I'm happy for there to not be a note on `CSub`; we'll set that bridge alight when/if we come to it.

Reply:
You can have notes about this kind of stuff if you want to but please keep it out of the main spec chapters (or tacked on at the end of them) because they're completely unexplored, likely to break various things throughout the software stack and add confusion. We really do not want people thinking it's a core part of CHERI, or seeing it and running with the idea before we've even figured out whether it makes any sense.

Reply:
I think what probably makes sense for the time being then perhaps is that we should have a section in the experimental chapter that talks about storing non-address bits in the low bits and what the impacts that would be in more of an architecture-neutral way.  That probably belongs as a separate thing from these changes however (and it would also be relevant for at least x86, so better to discuss it over there).